core/common/src/internal/annotations/FieldFormatter.ts

```diff
@@ -0,0 +1,160 @@
+  },
+
+  "boolean": (v, o) => {
+    const opts = o?.boolean;
+    if (typeof v !== "boolean" || !opts) {
+      return undefined;
+    }
+
+    return formatString(v ? opts.trueString : opts.falseString, o);
+  },
+
+  "quantity": (v, o) => formatString(formatQuantity(v, o?.quantity), o),
+
+  "coordinate": (v, o) => {
+    if (!isCoordinate(v)) {
+      return undefined;
+    }
+
+    const parts: string[] = [];
+    const formatComponent = (component: number) => {
+      const str = formatQuantity(component, o?.quantity);
+      if (undefined === str) {
+        return false;
+      }
+
+      parts.push(str);
+      return true;
+    };
+
+    const selector = o?.coordinate?.components ?? "XYZ";
+    if ((selector === "XYZ" || selector === "XY" || selector === "X") && !formatComponent(v.x)) {
+      return undefined;
+    }
+
+    if ((selector === "XYZ" || selector === "XY" || selector === "Y") && !formatComponent(v.y)) {
+      return undefined;
+    }
+
+    if (selector === "XYZ" && undefined !== v.z && !formatComponent(v.z)) {
+      return undefined;
+    }
+
+    const components = parts.join(o?.coordinate?.componentSeparator ?? ",");
+    return formatString(components, o);
+  },
+
+  "datetime": (v, o) => formatString(formatDateTime(v, o?.dateTime), o),
+};
+
+function formatString(s: string | undefined, o?: FieldFormatOptions): string | undefined {
+  if (undefined === s || !o) {
+    return s;
+  }
+
+  switch (o.case) {
+    case "upper":
+      s = s.toUpperCase();
+      break;
+    case "lower":
+      s = s.toLowerCase();
+      break;
+  }
+
+  if (o.prefix || o.suffix) {
+    s = `${o.prefix ?? ""}${s}${o.suffix ?? ""}`;
+  }
+
+  return s;
+}
+
+function formatEnum<T extends number | string>(v: T, o: EnumFieldFormatOptions<number | string> | undefined): string | undefined {
+  const fallback = o?.fallbackLabel;
+  const labels = o?.labels;
+  if (!labels) {
+    return fallback;
+  }
+
+  for (const entry of labels) {
+    if (entry.value === v) {
+      return entry.label;
+    }
+  }
+
+  return fallback;
+}
+
+function formatQuantity(v: FieldPrimitiveValue, _o?: QuantityFieldFormatOptions): string | undefined {
+  if (typeof v !== "number") {
+    return undefined;
+  }
+
+  // ###TODO apply quantity formatting...
+  if (_o && _o.formatProps){
+    const formatName = _o.formatProps.name ?? "defaultFormat";
+    const format = Format.createFromFullyResolvedJSON(formatName, _o.formatProps);
+    const formatterSpec = new FormatterSpec(format.name, format, _o?.unitConversions, _o?.sourceUnit);
+    return formatterSpec.applyFormatting(v);
+  }else
+    return v.toString();
+}
+
+function formatDateTime(v: FieldPrimitiveValue, _o?: DateTimeFieldFormatOptions): string | undefined {
+  if (typeof v === "string") {
+    const date = new Date(v);
+    if (!isNaN(date.getTime())) {
+      if (_o && _o.formatOptions){
+        const formatter = new Intl.DateTimeFormat(_o.locale, _o.formatOptions);
```

Comment:
I considered and rejected using `Intl.DateTimeFormat` because you are [not guaranteed](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DateTimeFormat/supportedLocalesOf) that the locale you want to use will be available for use. We need to produce consistently-formatted dates regardless of what system the code runs on.